src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,267 @@
+        range.start = {
+          value: lastRowKey,
+          inclusive: false,
+        };
+      } else {
+        // EndKey is read, remove this range
+        ranges.splice(index, 1);
+      }
+    }
+  }
+}
+
+/**
+ * Create a ReadRowsResumptionStrategy object to specify retry behaviour
+ *
+ * @class
+ * @param {ChunkTransformer} chunkTransformer A ChunkTransformer stream defined
+ * in chunktransformer.ts which is typically used for parsing chunked data from
+ * the server into a format ready for the user. The lastRowKey parameter of the
+ * chunkTransformer object is used for resumption logic to determine what keys
+ * and ranges should be included in the request for instance.
+ * @param {GetRowsOptions} options Options provided to createreadstream used for
+ * customizing the readRows call.
+ * @param {TableStrategyInfo} tableStrategyInfo Data passed about the table
+ * that is necessary for the readRows request.
+ *
+ * @example
+ * ```
+ * const strategy = new ReadRowsResumptionStrategy(
+ *   chunkTransformer,
+ *   options,
+ *   {tableName: 'projects/my-project/instances/my-instance/tables/my-table'}
+ * )
+ * gaxOpts.retry = strategy.toRetryOptions(gaxOpts);
+ * ```
+ */
+export class ReadRowsResumptionStrategy {
+  private chunkTransformer: ChunkTransformer;
+  private rowKeys: string[];
+  private ranges: PrefixRange[];
+  private rowsLimit: number;
+  private hasLimit: boolean;
+  private options: GetRowsOptions;
+  private tableStrategyInfo: TableStrategyInfo;
+  rowsRead = 0;
+  constructor(
+    chunkTransformer: ChunkTransformer,
+    options: GetRowsOptions,
+    tableStrategyInfo: TableStrategyInfo
+  ) {
+    this.chunkTransformer = chunkTransformer;
+    this.options = options;
+    this.rowKeys = options.keys || [];
+    this.ranges = TableUtils.getRanges(options);
+    this.rowsLimit = options.limit || 0;
+    this.hasLimit = this.rowsLimit !== 0;
+    this.rowsRead = 0;
+    this.tableStrategyInfo = tableStrategyInfo;
+    // If rowKeys and ranges are both empty, the request is a full table scan.
+    // Add an empty range to simplify the resumption logic.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      this.ranges.push({});
+    }
+  }
+
+  /**
+   * Gets the next readrows request.
+   *
+   * This function computes the next readRows request that will be sent to the
+   * server. Based on the last row key calculated by data already passed through
+   * the chunk transformer, the set of row keys and row ranges is calculated and
+   * updated. The calculated row keys and ranges are used along with other
+   * properties provided by the user like limits and filters to compute and
+   * return a request that will be used in the next read rows call.
+   *
+   * @return {protos.google.bigtable.v2.IReadRowsRequest} The request options
+   * for the next readrows request.
+   */
+  getResumeRequest(): protos.google.bigtable.v2.IReadRowsRequest {
```

Comment:
nit, I think this should be called getResumptionRequest. I'm open minded about it though if you feel strongly otherwise

Reply:
In Java it is called [getResumeRequest](https://github.com/googleapis/java-bigtable/blob/b110f4307a67801f7a29fc1e2accab86f651ca8a/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java#L81) so that is why I gave it this name, but feel free to push back if you feel strongly about this. Note that this is internal so users will never see this name so I don't feel strongly about the name either way.

Reply:
that's a good reason to keep it as is. I don't mind if it means consistency within your team.